### ISSUE TITLE
Type narrow when using `Self` isinstance checking

### ISSUE THREAD

**Issue by yanyongyu** [microsoft/pyright#6088]
**Describe the bug**

The var with type `<some other type> | Self` is not narrowed when using `isinstance(var, cls)`.

**Code or Screenshots**

The type of the var in `else` block is not narrowed.

```python
from typing import Self

class A:
    @classmethod
    def test(cls: type[Self], id: int | Self):
        if isinstance(id, cls):
            reveal_type(id)  # Type of "id" is "Self@A"
        else:
            reveal_type(id)  # Type of "id" is "int | Self@A"
```

[Pyright playground perma link](https://pyright-playground.decorator-factory.su/?gzip=H4sIANiSHmUC_32MSwoCMRBE9zlFLWdgTpCVnsGliISkow35DOlGEDy8SQRx5dt1V72KrWboc-dyA-e9NsWJUjTGJyeCozXoHOaVSe81zEegCCXRxSexw6fz0C4bOFhwUbzmzvrxBxzBwkXUFU8Lhw3d_ckHjR7k0nXs9cb6zSgJ_W--AVdiYofIAAAA)

**Comment by erictraut** [comment]
I don't know if I'd consider this a bug. I agree that pyright is being more conservative than necessary when narrowing in this case. It's not specific to `Self`. It applies to any type variable. (`Self` is effectively a type variable that has an upper bound defined by its containing class.)

So, here's a more general version of the above code that doesn't use `Self` but instead uses a type variable `T`.
```python
T = TypeVar("T")

class A:
    @classmethod
    def test(cls: type[T], id: int | T):
        if isinstance(id, cls):
            reveal_type(id) # int
        else:
            reveal_type(id) # int | T
```

If we replace `T` with a concrete type, then the current narrowing behavior is required for type safety.

```python
class A:
    @classmethod
    def test(cls: type[A], id: int | A):
        if isinstance(id, cls):
            reveal_type(id)
        else:
            reveal_type(id) # Must be "int | A" for type safety
```

But in the specific case where we're using a type variable, it would be safe to further narrow the type in the negative (else) case.

I'll note that pyright's current behavior for `Self` and for `T` matches mypy's behavior in this case. So mypy is also more conservative than necessary in this case.

I'm going to change this from a bug to an enhancement request and consider it in that light.

**Comment by erictraut** [comment]
This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.331, which I just published. It will also be included in a future release of pylance.
